create basic site24x7 API client

martinohmann opened the pull request.

martinohmann committed:
implement generic api client and specific monitors endpoint

martinohmann committed:
improve api errors, add docs

martinohmann committed:
separate rest client from api, move endpoints to separate pkg

martinohmann committed:
fix example

mmartins-bonial committed:
MonitorGroups Endpoint (#1)

* add monitor_groups endpoint

martinohmann committed:
add option to directly create an API client from options (#2)

martinohmann committed:
separate response into dedicated type (#3)

martinohmann committed:
add automatic backoff-retry for retryable requests (#4)

martinohmann committed:
implement location profiles endpoint (#6)

mmartins-bonial committed:
add notification_profiles endpoint (#9)

mmartins-bonial committed:
add threshold_profiles endpoint (#10)

* add threshold_profiles endpoint

mmartins-bonial committed:
add user_groups endpoint (#11)

martinohmann committed:
add endpoint test suite (#7)

mmartins-bonial committed:
add user_groups tests (#12)

martinohmann committed:
add golangci-lint, fix linting issues, update README.md (#13)

martinohmann committed:
fix usage example indentation

mmartins-bonial committed:
add location_profiles tests (#15)

* add location_profiles tests

mmartins-bonial committed:
add notifications_profiles tests (#16)

* add notifications_profiles tests

mmartins-bonial committed:
add threshold_profiles tests (#18)

mmartins-bonial committed:
add it_automation (#19)

* add it_automation

mmartins-bonial committed:
add it_automation tests (#20)

martinohmann committed:
remove Endpoint suffix (#21)

martinohmann committed:
add fake client with fake endpoints (#23)

This is useful for consumers of this package who want to stub out calls
to the site24x7 client in tests.

bobthebuilderberlin commented on api/endpoints/fake/it_automation.go:
```diff
@@ -0,0 +1,50 @@
+package fake
```
maybe instead of `fake`, use the more standard term stub?
```suggestion
package stub
```

martinohmann replied:
Until now, i've never seen any go project use the term `stub`, but most of the projects use `fake` (e.g. kubernetes). So i'd like to stick to `fake`.

bobthebuilderberlin commented on api/endpoints/fake/location_profiles.go:
```diff
@@ -0,0 +1,50 @@
+package fake
```
Hm I am not a go pro, but these CRUD stubs seems to be repetitive, just changing the type in each of the files.
Isn't there a way to DRY?

martinohmann replied:
Not really, as the types are implementing different interfaces. I planned to write a super simple code generator for the API endpoints and fakes in case we add more endpoints, as their structure is identical but only the types are different.

bobthebuilderberlin commented on rest/client.go:
```diff
@@ -0,0 +1,58 @@
+package rest
```
Not sure this abstraction is really needed, but ok.

bobthebuilderberlin commented on client.go:
```diff
@@ -1,5 +1,120 @@
+
+const (
+	// APIBaseURL is the base url of the Site24x7 API.
+	APIBaseURL = "https://www.site24x7.com/api"
```
Maybe makes sense to externalize to a config file?

martinohmann replied:
Why should this be in a config file? The API client is specific to Site24x7, so it does not make sense to make this configurable ;). If you really need to use a different URL, you can pass it as a parameter to the `rest.Client` (as it is done in the tests).

bobthebuilderberlin commented on client.go:
```diff
@@ -1,5 +1,120 @@
+}
+
+// New creates a new Site24x7 API Client with Config c.
+func New(c Config) Client {
```
Good.

bobthebuilderberlin commented on api/endpoints/location_profiles.go:
```diff
@@ -0,0 +1,83 @@
+package endpoints
```
Again, isn't there a way to DRY on the repeated CRUD pattern?
E.g.location_profiles.go and monitor_groups are basically the same file, just different types.
Would like to see the CRUD pattern abstracted.

martinohmann replied:
See above, this can be achieved by implementing a cod generator, which we did not do yet for the first version.

bobthebuilderberlin commented on backoff/retry.go:
```diff
@@ -0,0 +1,135 @@
+// DefaultRetryConfig is the default config for retrying http requests.
+var DefaultRetryConfig = RetryConfig{
+	MinWait:    1 * time.Second,
+	MaxWait:    30 * time.Second,
```
I would like to see these default values in a config file rather than in-code.

martinohmann replied:
Can you elaborate why this is necessary? Libraries usually do not (and should not) require any configuration files and therefore should ship their default configuration in code. Users are free to override the RetryConfig in the config they pass to `site24x7.New`.

bobthebuilderberlin replied:
Still, even if you do not need to provide the configuration as a user, it would still be nice to have configuration separated from code and shipped with the library. 
This way, all the configuration could be seen at once and is not distributed around the codebase.

martinohmann replied:
added documentation for the RetryConfig in the README

bobthebuilderberlin commented on oauth/oauth.go:
```diff
@@ -0,0 +1,59 @@
+const (
+	// TokenURL is the URL from where access tokens for the Site24x7 API are
+	// obtained.
+	TokenURL = "https://accounts.zoho.com/oauth/v2/token"
```
external config file?

bobthebuilderberlin commented on rest/request.go:
```diff
@@ -0,0 +1,147 @@
+	return r
+}
+
+func (r *Request) buildRawURL() string {
```
This still needs to be implemented by ourselves in 2019?

martinohmann replied:
what do you mean? This just concatenates strings.

bobthebuilderberlin replied:
Yeah, but isnt there a lib that builds REST-style URL's?

bobthebuilderberlin reviewed (commented):
In general looks ok.
I would really like to see a better abstraction of the CRUD patterns in stubs and actual client.

martinohmann committed:
update README.md and example, add retry defaults

martinohmann committed:
add .golangci.yml

martinohmann committed:
update README.md

martinohmann committed:
update example in README.md

martinohmann committed:
add error handler to backoff-retry (#24)

By default retryablehttp swallows error message. The errorHandler passes
the actual error up to the caller so if can be displayed or logged.

martinohmann committed:
set omitempty struct tag correctly (#25)

Especially for boolean values the omitempty tag should be absent. This
might need more tweaking in the future.
